### ISSUE TITLE
Evolving Normalization-Activation Layers

### ISSUE THREAD

**Issue by sile** [sile/readings#225]
URL
-----

https://arxiv.org/abs/2004.02967

Abstract
----------

> Normalization layers and activation functions are fundamental components in deep networks and typically co-locate with each other. Here we propose to design them using an automated approach. Instead of designing them separately, we unify them into a single tensor-to-tensor computation graph, and evolve its structure starting from basic mathematical functions. Examples of such mathematical functions are addition, multiplication and statistical moments. The use of low-level mathematical functions, in contrast to the use of high-level modules in mainstream NAS, leads to a highly sparse and large search space which can be challenging for search methods. To address the challenge, we develop efficient rejection protocols to quickly filter out candidate layers that do not work well. We also use multi-objective evolution to optimize each layer's performance across many architectures to prevent overfitting. Our method leads to the discovery of EvoNorms, a set of new normalization-activation layers with novel, and sometimes surprising structures that go beyond existing design patterns. For example, some EvoNorms do not assume that normalization and activation functions must be applied sequentially, nor need to center the feature maps, nor require explicit activation functions. Our experiments show that EvoNorms work well on image classification models including ResNets, MobileNets and EfficientNets but also transfer well to Mask R-CNN with FPN/SpineNet for instance segmentation and to BigGAN for image synthesis, outperforming BatchNorm and GroupNorm based layers in many cases.

Table of Contents
--------------------

- (1) Introduction
- (2) Related Works
- (3) Search Space
  - Layer Representation
  - Primitive Operations
  - Random Graph Generation
  - Observation 1: High sparsity of the search space
  - Observation 2: Weak generalization across architectures
- (4) Search Method
  - (4-1) Evolution
  - (4-2) Rejection protocols to address high sparsity of the search space
  - (4-3) Multi-architecture evaluation to promote strong generalization
- (5) Experiments
  - (5-1) Generalization across Image Classifiers
  - (5-2) Generalization to Instance Segmentation
  - (5-3) Generalization to GAN Training
  - (5-4) Intriguing Properties of EvoNorms
- (6) Conclusion
- (A) Code Snippets in TensorFlow
- (B) Proxy Task and Anchor Architectures
- (C) Implementation Details
- (D) Candidate layers without Reranking
- (E) Additional Results

BibTeX
------

```bibtex
@misc{liu2020evolving,
      title={Evolving Normalization-Activation Layers}, 
      author={Hanxiao Liu and Andrew Brock and Karen Simonyan and Quoc V. Le},
      year={2020},
      eprint={2004.02967},
      archivePrefix={arXiv},
      primaryClass={cs.LG}
}
```

See Also
----------

- GitHub: https://github.com/tensorflow/tpu/tree/master/models/official/resnet
